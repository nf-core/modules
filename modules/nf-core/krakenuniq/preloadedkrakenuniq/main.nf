process KRAKENUNIQ_PRELOADEDKRAKENUNIQ {
    tag "$meta.id"
    label 'process_high'

    conda (params.enable_conda ? "bioconda::krakenuniq=1.0.0" : null)
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/krakenuniq:1.0.0--pl5321h19e8d03_0':
        'quay.io/biocontainers/krakenuniq:1.0.0--pl5321h19e8d03_0' }"

    input:
    tuple val(meta), path(fastqs)
    path  db
    val ram_chunk_size
    val save_output_fastqs
    val report_file
    val save_output

    output:
    tuple val(meta), path('*.classified{.,_}*')     , optional:true, emit: classified_reads_fastq
    tuple val(meta), path('*.unclassified{.,_}*')   , optional:true, emit: unclassified_reads_fastq
    tuple val(meta), path('*classified.txt')        , optional:true, emit: classified_assignment
    tuple val(meta), path('*report.txt')                           , emit: report

    path "versions.yml"                                            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args ?: ''

    def classified   = meta.single_end ? '"\${PREFIX}.classified.fastq"'   : '"\${PREFIX}.classified#.fastq"'
    def unclassified = meta.single_end ? '"\${PREFIX}.unclassified.fastq"' : '"\${PREFIX}.unclassified#.fastq"'
    def classified_option = save_output_fastqs ? "--classified-out ${classified}" : ''
    def unclassified_option = save_output_fastqs ? "--unclassified-out ${unclassified}" : ''
    def output_option = save_output ? '--output "\${PREFIX}.krakenuniq.classified.txt"' : ''
    def report = report_file ? '--report-file "\${PREFIX}.krakenuniq.report.txt"' : ''
    def compress_reads_command = save_output_fastqs ? 'gzip --no-name *.fastq' : ''
    if (meta.single_end) {
        """
        krakenuniq \\
            --db $db \\
            --preload \\
            --preload-size $ram_chunk_size \\
            --threads $task.cpus \\
            $args

        strip_suffix() {
            local result=\$1
            # Strip any file extensions.
            echo "\${result%%.*}"
        }

        printf "%s\\n" ${fastqs} | while read FASTQ; do \\
            PREFIX="\$(strip_suffix "\${FASTQ}")"

            krakenuniq \\
                --db $db \\
                --threads $task.cpus \\
                $report \\
                $output_option \\
                $unclassified_option \\
                $classified_option \\
                $output_option \\
                $args2 \\
                "\${FASTQ}"
        done

        $compress_reads_command

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            krakenuniq: \$(echo \$(krakenuniq --version 2>&1) | sed 's/^.*KrakenUniq version //; s/ .*\$//')
        END_VERSIONS
        """
    } else {
        """
        krakenuniq \\
            --db $db \\
            --preload \\
            --preload-size $ram_chunk_size \\
            --threads $task.cpus \\
            $args

        strip_suffix() {
            local result
            read result
            # Strip any trailing dot or underscore.
            result="\${result%_}"
            echo "\${result%.}"
        }

        printf "%s %s\\n" ${fastqs} | while read FASTQ; do \\
            read -r -a FASTQ <<< "\${FASTQ}"
            PREFIX="\$(printf "%s\\n" "\${FASTQ[@]}" |  sed -e 'N;s/^\\(.*\\).*\\n\\1.*\$/\\1\\n\\1/;D' | strip_suffix)"

            krakenuniq \\
                --db $db \\
                --threads $task.cpus \\
                $report \\
                $output_option \\
                $unclassified_option \\
                $classified_option \\
                $output_option \\
                --paired \\
                $args2 \\
                "\${FASTQ[@]}"
        done

        $compress_reads_command

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            krakenuniq: \$(echo \$(krakenuniq --version 2>&1) | sed 's/^.*KrakenUniq version //; s/ .*\$//')
        END_VERSIONS
        """
    }

    stub:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args ?: ''

    def classified   = meta.single_end ? '"\${PREFIX}.classified.fastq"'   : '"\${PREFIX}.classified#.fastq"'
    def unclassified = meta.single_end ? '"\${PREFIX}.unclassified.fastq"' : '"\${PREFIX}.unclassified#.fastq"'
    def classified_option = save_output_fastqs ? "--classified-out ${classified}" : ''
    def unclassified_option = save_output_fastqs ? "--unclassified-out ${unclassified}" : ''
    def output_option = save_output ? '--output "\${PREFIX}.krakenuniq.classified.txt"' : ''
    def report = report_file ? '--report-file "\${PREFIX}.krakenuniq.report.txt"' : ''
    def compress_reads_command = save_output_fastqs ? 'gzip --no-name *.fastq' : ''
    if (meta.single_end) {
        """
        echo krakenuniq \\
            --db $db \\
            --preload \\
            --preload-size $ram_chunk_size \\
            --threads $task.cpus \\
            $args

        strip_suffix() {
            local result=\$1
            # Strip any file extensions.
            echo "\${result%%.*}"
        }

        printf "%s\\n" ${fastqs} | while read FASTQ; do \\
            echo "\${FASTQ}"
            PREFIX="\$(strip_suffix "\${FASTQ}")"
            echo "\${PREFIX}"

            echo krakenuniq \\
                --db $db \\
                --threads $task.cpus \\
                $report \\
                $output_option \\
                $unclassified_option \\
                $classified_option \\
                $output_option \\
                $args2 \\
                "\${FASTQ}"

            touch "\${PREFIX}.classified.fastq.gz"
            touch "\${PREFIX}.krakenuniq.classified.txt"
            touch "\${PREFIX}.krakenuniq.report.txt"
            touch "\${PREFIX}.unclassified.fastq.gz"
        done

        echo $compress_reads_command

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            krakenuniq: \$(echo \$(krakenuniq --version 2>&1) | sed 's/^.*KrakenUniq version //; s/ .*\$//')
        END_VERSIONS
        """
    } else {
        """
        echo krakenuniq \\
            --db $db \\
            --preload \\
            --preload-size $ram_chunk_size \\
            --threads $task.cpus \\
            $args

        strip_suffix() {
            local result
            read result
            # Strip any trailing dot or underscore.
            result="\${result%_}"
            echo "\${result%.}"
        }

        printf "%s %s\\n" ${fastqs} | while read FASTQ; do \\
            read -r -a FASTQ <<< "\${FASTQ}"
            echo "\${FASTQ[@]}"
            PREFIX="\$(printf "%s\\n" "\${FASTQ[@]}" |  sed -e 'N;s/^\\(.*\\).*\\n\\1.*\$/\\1\\n\\1/;D' | strip_suffix)"
            echo "\${PREFIX}"

            echo krakenuniq \\
                --db $db \\
                --threads $task.cpus \\
                $report \\
                $output_option \\
                $unclassified_option \\
                $classified_option \\
                $output_option \\
                --paired \\
                $args2 \\
                "\${FASTQ[@]}"

            touch "\${PREFIX}.classified_1.fastq.gz" "\${PREFIX}.classified_2.fastq.gz"
            touch "\${PREFIX}.krakenuniq.classified.txt"
            touch "\${PREFIX}.krakenuniq.report.txt"
            touch "\${PREFIX}.unclassified_1.fastq.gz" "\${PREFIX}.unclassified_2.fastq.gz"
        done

        echo $compress_reads_command

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            krakenuniq: \$(echo \$(krakenuniq --version 2>&1) | sed 's/^.*KrakenUniq version //; s/ .*\$//')
        END_VERSIONS
        """
    }
}
