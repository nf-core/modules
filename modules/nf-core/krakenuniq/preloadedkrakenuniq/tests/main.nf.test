nextflow_process {

    name "Test Process KRAKENUNIQ_PRELOADEDKRAKENUNIQ"
    script "../main.nf"
    process "KRAKENUNIQ_PRELOADEDKRAKENUNIQ"
    tag "modules"
    tag "modules_nfcore"
    tag "krakenuniq"
    tag "krakenuniq/preloadedkrakenuniq"

    test("sarscov2 - Illumina FASTQ single - stub-run") {
        options "-stub-run"

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [id:'test', single_end:true],
                    [
                        file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                    ]
                ]
                input[1] = []
                input[2] = '8GB'
                input[3] = true
                input[4] = true
                input[5] = true
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }

    }

    test("sarscov2 - Illumina FASTQ paired-end - stub-run") {
        options "-stub-run"

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [id:'test', single_end:false],
                    [
                        file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                        file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true)
                    ]
                ]
                input[1] = []
                input[2] = '8GB'
                input[3] = true
                input[4] = true
                input[5] = true
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }

    }
}
