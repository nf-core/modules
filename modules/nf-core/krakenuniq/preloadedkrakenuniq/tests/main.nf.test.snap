{
    "fastq-single-stub": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_interleaved.classified.fastq.gz:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_interleaved.unclassified.fastq.gz:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_interleaved.krakenuniq.classified.txt:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                "versions.yml:md5,6abf6c733f53fa3b6aaaa6f06864ef0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-06T11:21:36.338887437"
    },
    "fastq-single": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_interleaved.classified.fastq.gz:md5,3bd95021a8fbced1be8039b990b28176"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_interleaved.unclassified.fastq.gz:md5,143c7eb70ca93cc2d5ea98767c370424"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_interleaved.krakenuniq.classified.txt:md5,88a734a9a9216cb0770a77f36c9f4e78"
                ]
            ],
            [
                "versions.yml:md5,6abf6c733f53fa3b6aaaa6f06864ef0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-06T11:17:43.586414914"
    },
    "fastq-paired": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.merged.classified.fastq.gz:md5,dd7651837cce63e6108e28f4f019aedb"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.krakenuniq.classified.txt:md5,ed5e19c7a88312cc04e483ac5f2579cd"
                ]
            ],
            [
                "versions.yml:md5,6abf6c733f53fa3b6aaaa6f06864ef0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-06T11:37:46.718293365"
    },
    "fasta-stub": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "genome.classified.fasta.gz:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "genome.unclassified.fasta.gz:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "genome.krakenuniq.classified.txt:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                "versions.yml:md5,6abf6c733f53fa3b6aaaa6f06864ef0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-06T11:28:27.729550991"
    },
    "fastq-paired-stub": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.merged.classified.fastq.gz:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.merged.unclassified.fastq.gz:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.krakenuniq.classified.txt:md5,a5704c35e6b573a45e3a344768fe6975"
                ]
            ],
            [
                "versions.yml:md5,6abf6c733f53fa3b6aaaa6f06864ef0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-05T20:06:20.262529457"
    },
    "fasta": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "genome.classified.fasta.gz:md5,e73599798195a519ba2565c3f0275b93"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "genome.krakenuniq.classified.txt:md5,8aafacd89a6aac98aaf512df0a7493d1"
                ]
            ],
            [
                "versions.yml:md5,6abf6c733f53fa3b6aaaa6f06864ef0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-06T11:36:00.24752418"
    }
}